File: DataDriven/expressJS/src/controllers/Post.js
const Post = require("../models/Post");
const Comment = require("../models/Comment");

exports.create = async (req, res) => {
  const postData = Object.assign({}, req.body.postData);
  try {
    const postCreation = await Post.create({ ...postData, id: 0 }).then(
      (postDataTemp) => postDataTemp.toObject()
    );
    if (Object.hasOwn(postCreation, "_id") && postCreation._id) {
      res.status(200).json({ success: true });
    } else {
      throw new Error("Error creating database entry!");
    }
  } catch (e) {
    res.status(500).json({ error: `${e}` });
    console.log(e);
  }
};

exports.get = async (req, res) => {
  const page = req.query.page || 1;
  const size = req.query.size || 10;
  try {
    const posts = await Post.find({})
      .populate("authors")
      .populate("categories")
      .skip(size * (page - 1))
      .limit(size)
      .exec();
    if (posts.length < 1) {
      res.status(200).json({ result: "No posts found." });
    } else {
      res.status(200).json({ result: posts });
    }
  } catch (e) {
    res.status(500).json({ error: `${e}` });
  }
};

exports.bulkStatusUpdate = async (req, res) => {
  const { postIDs, newStatus } = req.body;
  if (!newStatus || !Array.isArray(postIDs) || postIDs.length < 1)
    return res.status(422).json({ error: "Invalid request data." });
  try {
    const updated = await Post.updateMany(
      { id: { $in: postIDs } },
      { $set: { postStatus: newStatus } }
    );

    if (Object.hasOwn(updated, "acknowledged") && updated.acknowledged) {
      res.status(200).json({ success: true });
    } else {
      throw new Error("Error updating database entry!");
    }
  } catch (e) {
    res.status(500).json({ error: `${e}` });
    console.log(e);
  }
};

exports.edit = async (req, res) => {
  const postData = Object.assign({}, req.body.postData);
  if (Object.keys(postData).length < 1 || !Object.hasOwn(postData, "id"))
    return res.status(422).json({ error: "Invalid request." });
  try {
    const updated = await Post.findOneAndUpdate({ id: postData.id }, postData, {
      new: true,
    });
    if (updated) {
      res.status(200).json({ success: true });
    } else {
      throw new Error("Error updating database entry!");
    }
  } catch (e) {
    res.status(500).json({ error: `${e}` });
  }
};

exports.search = async (req, res) => {
  try {
    const search = req.query.search;
    if (!search)
      return res.status(422).json({ error: "Invalid request data." });
    const posts = await Post.find({
      $or: [
        { title: { $regex: `.*${search}.*`, $options: "i" } },
        { content: { $regex: `.*${search}.*`, $options: "i" } },
      ],
    });
    if (posts.length < 1) {
      res.status(200).json({ result: "No posts found." });
    } else {
      res.status(200).json({ result: posts });
    }
  } catch (e) {
    res.status(500).json({ error: `${e}` });
  }
};

exports.sortByDate = async (req, res) => {
  try {
    const posts = await Post.find({}).sort({ lastUpdated: "desc" });
    if (posts.length < 1) {
      res.status(200).json({ result: "No posts found." });
    } else {
      res.status(200).json({ result: posts });
    }
  } catch (e) {
    res.status(500).json({ error: `${e}` });
  }
};

exports.popularPosts = async (req, res) => {
  try {
    const maxPosts = await Comment.aggregate([
      {
        $group: {
          _id: "$post",
          count: { $sum: 1 },
        },
      },
      { $sort: { count: -1 } },
      { $limit: 10 },
    ]);
    let postIDs = [];
    if (maxPosts.legnth < 1) {
      res.status(200).json({ result: "No posts found." });
    } else {
      postIDs = maxPosts.map(({ _id }) => _id);
    }
    const posts = await Post.find({ _id: { $in: postIDs } })
      .populate("authors")
      .populate("categories");
    if (posts.length < 1) {
      res.status(200).json({ result: "No posts found." });
    } else {
      res.status(200).json({ result: posts });
    }
  } catch (e) {
    res.status(500).json({ error: `${e}` });
  }
};

File: DataDriven/Gin/internal/server/controllers/post.go
package controllers

import (
	"net/http"
	"strconv"
	"time"

	"Gin/internal/models"

	"github.com/gin-gonic/gin"
	"github.com/gin-gonic/gin/binding"
)

type PostInputDTO struct {
	PostData struct {
		Title       string `json:"title"`
		Slug        string `json:"slug"`
		Content     string `json:"content"`
		PostStatus  string `json:"postStatus"`
		LastUpdated string `json:"lastUpdated"`
		Authors     []uint `json:"authors"`
		Categories  []uint `json:"categories"`
	} `json:"postData"`
}
type PostUpdateDTO struct {
	PostData struct {
		ID          uint   `json:"id"`
		Title       string `json:"title"`
		Slug        string `json:"slug"`
		Content     string `json:"content"`
		PostStatus  string `json:"postStatus"`
		LastUpdated string `json:"lastUpdated"`
		Authors     []uint `json:"authors"`
		Categories  []uint `json:"categories"`
	} `json:"postData"`
}

func (s *Server) GetPostsHandler(c *gin.Context) {
	page, err := strconv.Atoi(c.DefaultQuery("page", "1"))
	if err != nil || page < 1 {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid page number."})
		return
	}

	size, err := strconv.Atoi(c.DefaultQuery("size", "10"))
	if err != nil || size < 1 {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid size number."})
		return
	}

	var posts []models.Post
	offset := size * (page - 1)
	if err := s.db.Preload("Authors").Preload("Categories").Offset(offset).Limit(size).Find(&posts).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	if len(posts) < 1 {
		c.JSON(http.StatusOK, gin.H{"result": "No posts found."})
	} else {
		c.JSON(http.StatusOK, gin.H{"result": posts})
	}
}

func (s *Server) CreatePostHandler(c *gin.Context) {
	var postData PostInputDTO
	if err := c.ShouldBindBodyWith(&postData, binding.JSON); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}
	var authors []models.Author
	if err := s.db.Where("id IN ?", postData.PostData.Authors).Find(&authors).Error; err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "failed to retrieve authors: " + err.Error()})
		return
	}
	var categories []models.Category
	if err := s.db.Where("id IN ?", postData.PostData.Categories).Find(&categories).Error; err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "failed to retrieve categories: " + err.Error()})
		return
	}
	lastUpdated, err := time.Parse("2006-01-02", postData.PostData.LastUpdated)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "invalid date format for LastUpdated"})
		return
	}

	postNew := models.Post{
		Slug:        postData.PostData.Slug,
		Title:       postData.PostData.Title,
		Content:     postData.PostData.Content,
		LastUpdated: lastUpdated,
		PostStatus:  postData.PostData.PostStatus,
		Categories:  categories,
		Authors:     authors,
	}
	if err := s.db.Create(&postNew).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, gin.H{"success": true})
}

func (s *Server) BulkStatusUpdateHandler(c *gin.Context) {
	var req struct {
		PostIDs   []uint `json:"postIDs"`
		NewStatus string `json:"newStatus"`
	}

	if err := c.BindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	if len(req.PostIDs) < 1 || req.NewStatus == "" {
		c.JSON(http.StatusUnprocessableEntity, gin.H{"error": "Invalid request data."})
		return
	}

	if err := s.db.Model(&models.Post{}).Where("id IN (?)", req.PostIDs).Update("post_status", req.NewStatus).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, gin.H{"success": true})
}

func (s *Server) EditPostHandler(c *gin.Context) {
	var postData PostUpdateDTO
	if err := c.ShouldBindBodyWith(&postData, binding.JSON); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	if postData.PostData.ID == 0 {
		c.JSON(http.StatusUnprocessableEntity, gin.H{"error": "Invalid request."})
		return
	}
	var lastUpdatedTime *time.Time
	if postData.PostData.LastUpdated != "" {
		parsedTime, err := time.Parse("2006-01-02 15:04:05", postData.PostData.LastUpdated)
		if err != nil {
			c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid date format for LastUpdated. Expected format: YYYY-MM-DD HH:MM:SS"})
			return
		}
		lastUpdatedTime = &parsedTime
	}

	updateData := map[string]interface{}{
		"title":       postData.PostData.Title,
		"slug":        postData.PostData.Slug,
		"content":     postData.PostData.Content,
		"post_status": postData.PostData.PostStatus,
	}

	if lastUpdatedTime != nil {
		updateData["last_updated"] = *lastUpdatedTime
	}

	if err := s.db.Model(&models.Post{}).Where("id = ?", postData.PostData.ID).Updates(updateData).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, gin.H{"success": true})
}

func (s *Server) SearchPostsHandler(c *gin.Context) {
	search := c.Query("search")
	if search == "" {
		c.JSON(http.StatusUnprocessableEntity, gin.H{"error": "Invalid request data."})
		return
	}

	var posts []models.Post
	if err := s.db.Where("title LIKE ? OR content LIKE ?", "%"+search+"%", "%"+search+"%").Preload("Authors").Find(&posts).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	if len(posts) < 1 {
		c.JSON(http.StatusOK, gin.H{"result": "No posts found."})
	} else {
		c.JSON(http.StatusOK, gin.H{"result": posts})
	}
}

func (s *Server) SortPostsByDateHandler(c *gin.Context) {
	var posts []models.Post
	if err := s.db.Order("last_updated desc").Preload("Authors").Find(&posts).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	if len(posts) < 1 {
		c.JSON(http.StatusOK, gin.H{"result": "No posts found."})
	} else {
		c.JSON(http.StatusOK, gin.H{"result": posts})
	}
}

func (s *Server) PopularPostsHandler(c *gin.Context) {
	var maxPosts []struct {
		PostID uint
		Count  int
	}

	if err := s.db.Table("comments").Select("post_id, count(*) as count").Preload("Authors").Group("post_id").Order("count desc").Limit(10).Scan(&maxPosts).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}
	if len(maxPosts) < 1 {
		c.JSON(http.StatusOK, gin.H{"result": "No posts found."})
		return
	}

	var postIDs []uint
	for _, maxPost := range maxPosts {
		postIDs = append(postIDs, maxPost.PostID)
	}

	var posts []models.Post
	for _, postID := range postIDs {
		var post models.Post
		if err := s.db.Where("id = ?", postID).First(&post).Error; err != nil {
			c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
			return
		}
		posts = append(posts, post)
	}

	if len(posts) < 1 {
		c.JSON(http.StatusOK, gin.H{"result": "No posts found."})
	} else {
		c.JSON(http.StatusOK, gin.H{"result": posts})
	}
}
